### PR TITLE
feat: support credentials from Google OAuth Client

### PR DESCRIPTION
**Pull request by dirkluijk**
## Change Description

Instead of assuming every credential is a `HttpCredentialsAdapter`, also support `Credential` and throw error otherwise.

## Checklist

- [X] Make sure to open an issue as a 
  [bug/issue](https://github.com/GoogleCloudPlatform//issues/new/choose) 
  before writing your code!  That way we can discuss the change, evaluate 
  designs, and agree on the general idea.
- [ ] Ensure the tests and linter pass
- [ ] Appropriate documentation is updated (if necessary)

## Relevant issues:

- Fixes https://github.com/GoogleCloudPlatform/cloud-sql-jdbc-socket-factory/issues/1093

### CODE REVIEW

**Comment by JoeWang1127** [comment]
@shubha-rajan could you review this PR?

**Comment by shubha-rajan** [comment]
Hi @dirkluijk thanks for opening this PR and thanks for adding that test! I'll take a look shortly.

File: core/src/main/java/com/google/cloud/sql/core/CloudSqlInstance.java [hunk 1]
```diff
@@ -169,6 +172,32 @@ class CloudSqlInstance {
     }
   }
 
+  private OAuth2Credentials parseCredentials(HttpRequestInitializer source) {
```
Comment (shubha-rajan) [inline core/src/main/java/com/google/cloud/sql/core/CloudSqlInstance.java:175]: Can you add a simple unit test for this method which tests all 3 cases (HttpCredentialsAdapter, Credential, and other unsupported type)

File: core/src/main/java/com/google/cloud/sql/core/CloudSqlInstance.java [hunk 2]
```diff
@@ -169,6 +172,32 @@ class CloudSqlInstance {
...
+      Credential credential = (Credential) source;
+      AccessToken accessToken = new AccessToken(
+          credential.getAccessToken(),
+          new Date(credential.getExpirationTimeMilliseconds())
+      );
+      GoogleCredentials googleCredentials = new GoogleCredentials(accessToken) {
+        @Override
+        public AccessToken refreshAccessToken() throws IOException {
+          credential.refreshToken();
+          return new AccessToken(credential.getAccessToken(), new Date(credential.getExpirationTimeMilliseconds()));
```
Comment (shubha-rajan) [inline core/src/main/java/com/google/cloud/sql/core/CloudSqlInstance.java:191]: Looks like the linter is failing on this line:
```suggestion
          return new AccessToken(
               credential.getAccessToken(),
               new Date(credential.getExpirationTimeMilliseconds())
          );
```

**Review by shubha-rajan — CHANGES_REQUESTED** [review]
Looks good so far. Have you tested this with real credentials and a real instance?

File: core/src/main/java/com/google/cloud/sql/core/CloudSqlInstance.java (hunk 1 above)
Comment (dirkluijk) [reply core/src/main/java/com/google/cloud/sql/core/CloudSqlInstance.java:175]: Will have a look at it.

File: core/src/main/java/com/google/cloud/sql/core/CloudSqlInstance.java (hunk 2 above)
Comment (dirkluijk) [reply core/src/main/java/com/google/cloud/sql/core/CloudSqlInstance.java:191]: Will fix it

**Comment by dirkluijk** [comment]
> Looks good so far. Have you tested this with real credentials and a real instance?
Yes, worked fine on [my example setup](https://github.com/dirkluijk/spring-gcp-sql-issue) when I tested it locally.

I'll have a look at another unit test + lint fix today.

**Comment by dirkluijk** [comment]
Fixed the lint issue and added another unit test.

**Comment by shubha-rajan** [comment]
@dirkluijk looks like the the lint check is failing again:
image
can you add this dependency to `core/pom.xml`? Once you do that I'm happy to approve and merge this.

**Comment by dirkluijk** [comment]
@shubha-rajan  Sorry, I thought nothing changed it terms of dependencies. Fixed it!

**Review by shubha-rajan — APPROVED** [review]

**Pull request merged**